lib/clients.js

```diff
@@ -0,0 +1,99 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const error = require('./error')
+const Pool = require('./pool')
+
+/**
+ * @typedef {Object} ClientInstance
+ * @property {string} id Client Instance ID
+ * @property {string} clientId
+ * @property {string} pushEndpoint
+ * @property {string} pushPublicKey
+ * @property {string} pushAuthKey
+ * @property {Object} availableCommands
+ */
+module.exports = (log, db, config) => {
+  const SafeUrl = require('./safe-url')(log)
+  const path = new SafeUrl('/v1/refresh-tokens/:uid')
+  const pool = new Pool(config.oauth.url, { timeout: 15000 })
+  const headers = {Authorization: `FxA-Shared-Secret ${config.oauth.sharedSecretKey}`}
+  return {
+    /**
+     * @param {String} uid
+     * @returns {ClientInstance[]}
+     */
+    async getClientsInstances(uid) {
+      const params = {uid: uid.toString('hex')}
+      // Refresh tokens are the clients list authority, whether we have
+      // stored metadata or not is irrelevant.
+      const refreshTokens = await pool.get(path, params, {headers})
+      const knownInstances = await db.clientsInstances(uid)
+      const knownInstancesById = new Map(knownInstances.map(instance => [instance.id, instance]))
+      // We use a map to avoid having duplicate instanceIDs
+      // (multiple refresh tokens can have the same instance ID).
+      const instancesById = new Map(refreshTokens.map(({ clientId, instanceId }) => {
+        const metadata = knownInstancesById.get(instanceId) || {
+          id: instanceId,
+          clientId,
+          name: null,
+          pushEndpoint: null,
+          pushPublicKey: null,
+          pushAuthKey: null,
+          availableCommands: {},
+        }
+        return [instanceId, metadata]
+      }))
+      return [...instancesById.values()]
+    },
+
+    /**
+     * Retrieves all the device records AND the client instances records in an account
+     * for use with Push.
+     * @param {String} uid
+     * @returns {PushTarget[]}
+     */
+    async accountPushClientsAndDevices(uid) {
+      const [devices, clientsInstances] = await Promise.all([db.devices(uid), this.getClientsInstances(uid)])
```

Comment:
This method replaces some usage of `request.app.devices` and it would be preferable to use `request.app.devices` in here too where possible, because that guards against hitting the db multiple times per request. The devices query is not one of our faster ones and, for instance, we have metrics code that hits it regularly too (every time a flow event or amplitude event is emitted). If all the places that hit it go through `request.app.devices` it means we'll only call it once per request, reducing load on the db.

Of course, there are legitimate occasions where you can't use the cached devices list, if you don't have a request object or you know the device list has been modified. But we'd want to try and limit it to just those specific cases and not blanket-ignoring `request.app.devices` for every call to this method, if possible.

So, concrete suggestion, can we add a `request` arg to this method, then use `request.app.devices` if it's set? Then somewhere like `lib/profile/updates.js` where you don't have the request object will fall back to calling `db.devices` whereas others will continue to use the cached copy. Or some other approach that achieves the same end result of course.